bluetooth-mitm/source/controllers/dualsense_controller.cpp

```diff
@@ -22,6 +22,47 @@ namespace ams::controller {
 
         const constexpr float stick_scale_factor = float(UINT12_MAX) / UINT8_MAX;
 
+        const uint8_t player_led_flags[] = {
+            0x04,
+            0x0A,
+            0x15,
+            0x1B,
+            0x1F
+        };
+
+        const RGBColour player_led_colours[] = {
+            // Same colours used by PS4
+            {0x00, 0x00, 0x40}, // blue
+            {0x40, 0x00, 0x00}, // red
+            {0x00, 0x40, 0x00}, // green
+            {0x20, 0x00, 0x20}, // pink
+            // New colours for controllers 5-8
+            {0x00, 0x20, 0x20}, // cyan
+            {0x30, 0x10, 0x00}, // orange
+            {0x20, 0x20, 0x00}, // yellow
+            {0x10, 0x00, 0x30}  // purple
+        };
+
+    }
+
+    Result DualsenseController::Initialize(void) {
+        R_TRY(EmulatedSwitchController::Initialize());
+        R_TRY(this->PushRumbleLedState());
+
+        return ams::ResultSuccess();
+    }
+
+    Result DualsenseController::SetPlayerLed(uint8_t led_mask) {
+        uint8_t player_number;
+        R_TRY(LedsMaskToPlayerNumber(led_mask, &player_number));
+        m_led_flags = player_led_flags[std::min(player_number, (uint8_t)4)];
```

Comment:
What are the led flags used for? Do we want the same value for players 5-8?

Also, that `uint8_t` cast probably isn't necessary. If there's a reason for it, consider using c++ style casts for consistency with the rest of the project.

Reply:
The LED flags enable the player indicator lights along the bottom of the touchpad:
![player-leds](https://user-images.githubusercontent.com/3057954/106011237-44249900-6088-11eb-994f-a968eb48b59a.png)
There are only 5 LEDs total, so for players 5-8 I just set them all to be enabled.

Regarding the cast, the compiler complains that the template isn't satisfied without it. I'll convert it to C++ style, my bad :P

Reply:
Oh I see. Can the LEDs be addressed individually or do you only have the above 4 options? The Switch uses combinations of 2+ LEDs for players 5-8, so we could also do that here if the controller allows.

If the compiler is complaining it might be due to an ambiguous return type. You can probably avoid the cast and force it like this `std::min<uint8_t>(player_number, 4)`

Reply:
Ah, I wasn't aware of how the Switch handles it; that should be no problem.

Reply:
You can see the bit patterns used by the switch for each player here. Note that the LSB represents the leftmost LED

https://github.com/ndeadly/MissionControl/blob/a916172c07bf31e3af693f860e394af838aa0096/bluetooth-mitm/source/controllers/switch_controller.cpp#L22

Reply:
Thanks! I looked up a YouTube video to see them already, heh. I mimicked them all except for player 8, since that pattern is technically already taken by player 2 on the DualSense; instead I enabled the three middle LEDs for that one. It should be good to go now, unless you have any other suggestions :)

Reply:
Is it possible rather than centering the LED patterns, to always operate on the 4 LEDs shown in that last picture and left-align the patterns like the switch does? This would avoid the issue of the clashing patterns for players 2 and 8, and be more switch-like on the whole. Otherwise, everything looks good to me now too.

Reply:
It's certainly possible; the main reason I stuck with the PS5 patterns is because I found it looked a little strange to have them off-center when they're at pretty much the focal point of the controller. If you'd still prefer left-aligning them though, I'm happy to change it!

Reply:
I think I would prefer for them to match what is set on other Nintendo controllers for consistency's sake. I'm sure someone will inevitably complain to me that player 8 looks like player 3 with the current scheme. Unless you're not just speaking from personal taste and it actually looks bad left aligned for some reason?

I'm not too hung up on it either way though so will trust your judgement if you think the current scheme works :)

Reply:
The more I think about it, the more I agree with you. I also discovered while changing it that the leftmost LEDs are controlled by the least significant bits, when all this time I had assumed it was the other way around :P
Anyway, I've pushed the update, so everything should finally be good!